Ejabberd muc sub

reitzig opened the pull request:
Adds a new module that supports [ejabberd Muc Sub](https://docs.ejabberd.im/developer/xmpp-clients-bots/proposed-extensions/muc-sub/), a non-XEP feature.

Commit:
HSDKA-178 Defined public interface of module and delegate.

Commit:
HSDKA-178 Implemented all functions.

Commit:
HSDKA-178 Added MUC-Sub discovery.

Commit:
HSDKA-178 nonnull/nullable treatment

Commit:
HSDKA-178 Coding style.

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+@class XMPPRoom;
+
+/**
+ * The XMPPMUCSub provides functionality to a proprietary Multi User Chat extension of 
```
"provides ... to" --> "provides support for"

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+/**
+ * The XMPPMUCSub provides functionality to a proprietary Multi User Chat extension of 
+ * the ejabberd XMPP server. This extension aims to provide a solution to the problem
+ * that users are required to send a presence to a MUC room in oder to receive messages.
```
*order

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+ * By subscribing to the room a user can also participate if not online. Once reconnected,
+ * missed messages are synced.
+ * 
+ * The extension leverages several existing extension to achieve its task. More details
```
*extensions

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+ *        The `XMPPRoom` for which to check if MUC-Sub has been enabled.
+ * 
+ * @return
+ * The request id of the IQ is in case client code may want to do manual tracking. `nil`
```
Clarify that the return value is indeed the XMPPElement ID, that is nothing special? (applies to all instances)

Digithurst-RL replied:
"Returns the randomly generated "id" attribute value of the standard <iq> element that 
is sent in case client code may want to do manual tracking."?

reitzig replied:
If there's no precedence in XMPPFramework, let's stick to the wording in [the RFC](https://xmpp.org/rfcs/rfc6120.html#stanzas-attributes-id):

"'id' attribute of the IQ stanza that was sent to the server" and similar

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+ * @param nick
+ *        Ones nickname in the room. If `nil`, the bare JID is used.
+ * 
+ * @param pass
```
*password

Digithurst-RL replied:
No, it's pass. The value after : is used.

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+ * if `room` is `nil`.
+**/
+- (nullable NSString *)subscribeTo:(nonnull XMPPJID *)room nick:(nullable NSString *)nick 
+                          password:(nullable NSString *)pass;
```
Note to self: unfortunately, there does not seem a way to write methods like this so that the bridged Swift interface uses default values.

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+ *        The room's JID to which `user` subscribes to.
+ * 
+ * @param nick
+ *        Ones nickname in the room. If `nil`, the bare JID is used.
```
Of the current user or `user`?

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+ * The request id of the IQ is returned in case client code may want to do manual 
+ * tracking.
+**/
+- (nullable NSString *)subscribersIn:(nonnull XMPPJID *)room;
```
The English phrase would be "list the subscribers of [room]", so `subscribersOf(...)`?

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+
+
+/**
+ * The user has been unsubscribed to a specific room. It is not differentiated between 
```
*from a specific room

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.h:
```diff
@@ -0,0 +1,272 @@
+**/
+- (void)xmppMUCSubDidFailToReceiveSubscribersIn:(nonnull XMPPMUCSub *)sender 
+                                             to:(nonnull XMPPJID *)room 
+                                          error:(nonnull NSError *)error;
```
Should the signature here be "split"? (You know what I mean, hopefully. :'D)

Digithurst-RL replied:
`- (void)xmppMUCSub:(nonnull XMPPMUCSub *)sender didFailToReceiveSubscribersOf:(nonnull XMPPJID *)room 
             error:(nonnull NSError *)error;`

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.m:
```diff
@@ -0,0 +1,699 @@
+
+- (id)init
+{
+    return [self initWithDispatchQueue:nil];
```
XMPPFramework uses `NULL`. `nil` seems to be used for Objective-C objects; is `dispatch_queue_t` an Objective-C or C class?

If I understand [this](https://developer.apple.com/documentation/dispatch/dispatch_queue_t) correctly, it's the former and `nil` is correct.

Digithurst-RL replied:
It's a type alias to an Objective-C class, so `nil` is correct.

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.m:
```diff
@@ -0,0 +1,699 @@
+    NSString *iqId = [XMPPStream generateUUID];
+    
+    XMPPIQ *iq = [XMPPIQ iqWithType:@"get" elementID:iqId];
+    [iq addAttributeWithName:@"from" stringValue:xmppStream.myJID.bare];
```
Why drop the resource here?

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.m:
```diff
@@ -0,0 +1,699 @@
+    [iq addChild:query];
+    
+    [xmppIDTracker addElement:iq target:self selector:@selector(handleSupportedByIQ:withInfo:) 
+                      timeout:60];
```
What does XMPPIDTracker do with this? Were will the response go? (I'm sure it's down below somewhere, but I think a comment here would be helpful.)

Also, shouldn't the timeout be configurable, or use a timeout configured in XMPPFramework?

(Applies to several similar instances below.)

reitzig replied:
Ohh, it calls the method given by `@selector`! Neat, never mind.

Digithurst-RL replied:
The only defined timeout I could find was no timeout, waiting indefinitely. Other code I have seen used the hardcoded 60 seconds.

reitzig replied:
Oh my ... it _is_ an ... interesting framework.

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.m:
```diff
@@ -0,0 +1,699 @@
+
+- (NSString *)subscribeTo:(XMPPJID *)room nick:(NSString *)nick password:(NSString *)pass
+{
+    return [self subscribe:xmppStream.myJID to:room nick:nick password:pass];
```
What does `xmppStream.myJID` yield if the stream is not active/online?
What do our methods do if the stream is not active/online?

--> Do we need to guard against that?

(Applies to all (?) methods.)

Digithurst-RL replied:
Good point. None of the code I've looked at did this so I totally forgot about that myself.

reitzig replied:
Oh my, it _is_ an ... interesting framework.

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.m:
```diff
@@ -0,0 +1,699 @@
+    // <iq from='hag66@shakespeare.example/pda'
+    //       to='coven@muc.shakespeare.example'
+    //     type='get'
+    //       id='ik3vs715'>
```
`ik3vs715` is not a UUID. ;)

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.m:
```diff
@@ -0,0 +1,699 @@
+        }
+    }};
+    
+    if (dispatch_get_specific(moduleQueueTag)) {
```
What does this do? `true` if the current queue has label `moduleQueueTag`?

If so, not sure if executing `block` synchronously is the right thing to do. Older incoming "events" may be waiting on the queue?

Digithurst-RL replied:
I copied this pattern from other modules. I noticed that XMPPIDTracker asserts that it is called on the correct queue. Can't really explain this construct.

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.m:
```diff
@@ -0,0 +1,699 @@
+    if ([type isEqualToString:@"result"] || [type isEqualToString:@"error"]) {
+        return [xmppIDTracker invokeForID:iq.elementID withObject:iq];
+    }
+    
```
What does this do, delegate the decision whether this was an IQ response we waited for to `xmpppIDTracker`?

Digithurst-RL replied:
Yes.

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.m:
```diff
@@ -0,0 +1,699 @@
+        block();
+    }
+    else {
+        dispatch_async(moduleQueue, block);
```
moduleQueue is serial, right?

Digithurst-RL replied:
It's either the queue that is provided on initialization or the initializer of `XMPPModule` creates one.
 ```swift
const char *moduleQueueName = [[self moduleName] UTF8String];
moduleQueue = dispatch_queue_create(moduleQueueName, NULL);
```

reitzig commented on Extensions/MUC-Sub/XMPPMUCSub.m:
```diff
@@ -0,0 +1,699 @@
+}
+
+
+- (NSArray<XMPPJID *>*)jidFromSubscriptionIQ:(XMPPIQ *)iq
```
Whose JID?

Should headers/signatures for these helpers be above in `@interface XMPPMUCSub (PrivateAPI)`?

Digithurst-RL replied:
Depends on the context. This is the structure that is parsed. 
```xml
<iq from='muc.shakespeare.example'
    to='hag66@shakespeare.example'
    type='result'
    id='E6E10350-76CF-40C6-B91B-1EA08C332FC7'>
  <subscriptions xmlns='urn:xmpp:mucsub:0'>
    <subscription jid='coven@muc.shakespeare.example' />
    <subscription jid='chat@muc.shakespeare.example' />
  </subscriptions>
</iq>
```
The < subscriptions > tags are read and either mean subscribed users in a room or room a user is subscribed to.

I removed the `(PrivateAPI)` extension because it's not necessary.

Digithurst-RL replied:
Added structure as comment at the top of the function.

reitzig reviewed (commented):
We should document what happens if a subscribed user actually *joins* the room (using usual MUC as per XEP-0045). That is, repeat what the "specification" says.

Is it obvious that all the methods on XMPPMucSub are asynchronous, resp. when they are? Maybe a short note in the protocol documentation would be a good idea.

Found only small stuff. Tests will have to show that it actually works, of course.

Commit:
HSDKA-178 Code review. Added blocks in functions that use XMPPIDTracker.

Supposed to fix crash due to wrong queue being used ("invoked on incorrect queue").

Commit:
HSDKA-178 Wrong value for nick was used.

Commit:
HSDKA-178 Unsubscribe handler contained invalid assumption about response format.

Commit:
HSDKA-178 Message and presence were sent with wrong function.

Commit:
HSDKA-178 Presences event is plural.

ejabberd documentation of MUC-Sub has typo.

reitzig commented:
@Digithurst-RL  Looks good to me so far. Can we "merge"?
